Country category filter version

Chun-Lin opened the pull request:
Now News Filter can Filter news by 
1. A keyword
2. A country
3. A category

<img width="1437" alt="2018-06-25 4 55 01" src="https://user-images.githubusercontent.com/12542369/41840539-94f4c934-7898-11e8-800e-c4fc0f3150ad.png">


For more details, please check out README.md

Chun-Lin committed:
feat: add gh-pages packages

Chun-Lin committed:
style: add logo images and guideline colors

Chun-Lin committed:
style: add header styled

Chun-Lin committed:
style: add input style

Chun-Lin committed:
style: add search icon on the left of searchbar

Chun-Lin committed:
fix: fox some minor pixel adjustment

Chun-Lin committed:
style: add Logo and fix some layout inside the header

Chun-Lin committed:
refactor: modify the folder structure

Chun-Lin committed:
style: finish the location block and make a country selector component

Chun-Lin committed:
style: add Datepicker block in the sidebar

Chun-Lin committed:
style: finish the sidebar layout

Chun-Lin committed:
style: add resultpage layout and modify the sidebar folder structure, and remove some useless code

Chun-Lin committed:
style: add result title component

Chun-Lin committed:
style: add tag component to the ResultPage

Chun-Lin committed:
style: add result list component to result page

Chun-Lin committed:
fix: use the original datepicker and give it the style

Chun-Lin committed:
fix: give Sidebar title bold style

Chun-Lin committed:
style: give some minor changes to datepicker styles, and change it to functional comopnent

Chun-Lin committed:
style: add custom style to checkbox component and modify the layout of Category container

Chun-Lin committed:
style: add pagination styles and make result list can show more not just within the view height

Chun-Lin committed:
style: make the <main> background-color to LIGHT_GREY

Chun-Lin committed:
fix: fix HEADER component z-index, make the img src in ResultList to be a prop

Chun-Lin committed:
refactor: modify the structure of Result List to make it more clean

Chun-Lin committed:
refactor: renmae ResultTitle to ResultCount, and modify the css structure of ResultCount

Chun-Lin committed:
refactor: rename TagWrapper to TagLayout

Chun-Lin committed:
refactor: modify the structure of pagination, now can pass pageCount to pagination component to show

Chun-Lin committed:
style: modify the HTML instructure in App.js

Chun-Lin committed:
refactor: change SearchBar to Class component

Chun-Lin committed:
fix: give onChange event noop function and give key to Pagination

Chun-Lin committed:
fix: fix when the image is larger then 220x220, and the layout will be broken

Chun-Lin committed:
feat: use redux to fetch News API and show info in the resultList

now image, Title, Description, Author, Time can be displayed correct in the list. Lack of Category
and Location responses, so I'll add another info after

Chun-Lin committed:
feat: add results number to the resultspage

Chun-Lin committed:
fix: fix the author is empty when there is no author

give 'Anonymous' when the author response is empty

Chun-Lin committed:
fix: add alt messages when there is no image

Chun-Lin committed:
fix: make sidebar container's height to fit the whole page height

Chun-Lin committed:
style: hide pagination when there is no articles

Chun-Lin committed:
fix: change checkbox to radio button'

Chun-Lin committed:
feat: add location filter feature, and add Search botton for dispatching fetchNews action

now fetchNews is a action creator for fetching data via getState, and SearchBar, CountrySelector,
Category are update queryString in the store

Chun-Lin committed:
feat: add category filter feature, and remove Tag component in resultPage

Chun-Lin committed:
feat: add filterButton component

Chun-Lin committed:
docs: change the website title to News Filter

Chun-Lin committed:
style: finish the style of FilterButton

Chun-Lin committed:
feat: remove Search button and Filter button because now news can be fetched after fininshing input

Chun-Lin committed:
feat: add Loading Spinner to resultpage, Loading component will show up before receiving responses d

Chun-Lin committed:
feat: add QUERY_PAGE action types

Chun-Lin committed:
fix: fix the searchBar & resultPage layout are crash in Windows OS

Chun-Lin committed:
feat: now News Filter can choose page by clicking pagination

Chun-Lin committed:
refactor: remove useless fontAwesome icon

Chun-Lin committed:
refactor: remove useless console.log()

Chun-Lin committed:
feat: use immer to handle data manipulation in reducer and add fetchNewsFail action manipulation

Chun-Lin committed:
style: remove overflow manipulation in resultList, add overflow manipulation in Description, add anc

Chun-Lin committed:
refactor: refactor the components in the sideBar container

Chun-Lin committed:
refactor: refactor components in the Header container

Chun-Lin committed:
fix: fix the loading spinner won't show up and refactor the comopnents in resultPage

Chun-Lin committed:
refactor: refactor pagination components to make it more clean

Chun-Lin committed:
refactor: use redux-actions and immer to handle reducers, and split reducers to fetch and query

Chun-Lin committed:
fix: remove useless code in index.js

Chun-Lin committed:
refactor: shorten the return statement

Chun-Lin committed:
refactor: refactor ResultPage container

Chun-Lin committed:
fix: add key into Category component and fix the typo on PropTypes

Chun-Lin committed:
refactor: remove unused code and add missing proptypes

Chun-Lin committed:
refactor: revmoe images folder in assets

Chun-Lin committed:
refactor: move colors.js to constants

Chun-Lin committed:
fix: remove useless import in App.js

Chun-Lin committed:
Update README.md for fitting News Filter

Chun-Lin committed:
refactor: modify the READEME.md to make it more clean

purepennons commented on src/index.js:
```diff
@@ -1,8 +1,49 @@
+  query: queryStringReducer,
+})
+
+const store = createStore(rootReducer, composeEnhancers(applyMiddleware(thunk)))
```
有特別區分 store 資料夾了，createStore 這類設定應該在裡面才合理。

store 資料夾其實命名成 `redux` 相關的名稱較合理。store 中並不會有 actions，而是 redux 中才有。

purepennons commented on src/store/actions/fetchNews.js:
```diff
@@ -0,0 +1,48 @@
+
+  axios
+    .get(
+      `?apiKey=${API_KEY}&q=${
```
直接針對 querystring object 找個 qs module 一口氣 stringify 即可

purepennons replied:
而且這種寫法，如果沒有選項，會導致 qs 有 key 但沒有 value

purepennons commented on src/store/actions/fetchNews.js:
```diff
@@ -0,0 +1,48 @@
+        queryString.page
+      }`,
+    )
+    .then(res => {
```
es2018 有 `Promise.finally` 可以使用，所以其實關於 loading 狀態的處理，可以拆成獨立 store state，透過 finally 一口氣關閉 loading 狀態。而不用分別在 `fetchNewsSuccess ` 與 `fetchNewsFail ` 處理 loader 的關閉。

purepennons commented on src/store/actions/fetchNews.js:
```diff
@@ -0,0 +1,48 @@
+  articles: articles,
+})
+
+export const fetchNewsFail = err => ({
```
雖然沒有硬性要求，不過 action 的形式可以以 standard action 為主

purepennons commented on src/store/actions/fetchNews.js:
```diff
@@ -0,0 +1,48 @@
+  const { queryString } = getState().query
+
+  let countryValue = ''
+  queryString.country.value
```
這寫法太多冗餘的程式碼了。

```js
const countryValue = queryString.country.value? queryString.country.value.toLowerCase(): ''
```

purepennons commented on src/store/reducers/fetchNewsReducer.js:
```diff
@@ -0,0 +1,36 @@
+
+const fetchNewsReducer = handleActions(
+  {
+    [FETCH_NEWS_INIT]: produce(draft => {
```
1. 命名為 init，但實際上行為並不是初始化，loading 為 true
2. 如果將 loading 獨立處理，這邊只需要讓 draft 為 initState 即可

purepennons commented on src/store/actions/fetchNews.js:
```diff
@@ -0,0 +1,48 @@
+  error: err,
+})
+
+export const fetchNews = () => (dispatch, getState) => {
```
對於非同步的 action creator 若可以 return Promise，是個不錯的作法，讓你 component 若需要在生命週期做連續的非同步操作，保留操作的空間。

purepennons commented on src/store/actions/query.js:
```diff
@@ -0,0 +1,26 @@
+  QUERY_PAGE,
+} from './actionTypes'
+
+export const queryCountry = country => ({
```
讓 action creator 改成可以吃 key 跟 value 的形式，就一直寫重複的 action creator

```js
export const query = (k, v) => ({
  type: Query,
  payload: { [k]: v }
})
```

purepennons replied:
或者保留原本的寫法，但是 reducer 可以化簡成類似這樣：

```js
const reducer = (state, action) => {
  if(/^QUERY_/.test(action.type)) {
    return {...state, ...action.payload}
  } else {
    return state
  }
}
```

Chun-Lin committed:
style: wrap Logo, SearchBar component to styled-component, and now the render structure is normal ht

Chun-Lin committed:
refactor: use query-string module to stringified query string & use async/await

Chun-Lin committed:
refactor: add promise.finally to hide loading spinner

Chun-Lin committed:
refactor: move loading state to a indepedent state

Chun-Lin committed:
refactor: integrate all QUERY_XXX  actions to QUERY

now there is just one action called QUERY, you can inject QUERY type and value as parameters

Chun-Lin committed:
refactor: using duck pattern to style redux and containers

Chun-Lin committed:
refactor: refactor all comopnents in ResultPage using css selector within styled-components

Chun-Lin committed:
refactor: install query-string@5 & remove useless code

Chun-Lin committed:
refactor: refactor all styled-components in Sidebar to css selector within a styled-component

Chun-Lin committed:
feat: add Html tag to verify from google search console

Chun-Lin committed:
fix: fix the proptypes bug to match styled-components
